Infrastructure/6139 lodash imports

eugene-manuilov opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses in the following list. -->
Addresses issue:

- #6139

## Relevant technical choices

<!-- Please describe your changes. -->

## PR Author Checklist

- [ ] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 5.2 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [x] Run the code.
- [x] Ensure the acceptance criteria are satisfied.
- [x] Reassess the implementation with the IB.
- [x] Ensure no unrelated changes are included.
- [x] Ensure CI checks pass.
- [x] Check Storybook where applicable.
- [x] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

techanvil commented on assets/js/googlesitekit/modules/datastore/modules.js:
```diff
@@ -20,18 +20,16 @@
-import merge from 'lodash/merge';
-import isPlainObject from 'lodash/isPlainObject';
 import invariant from 'invariant';
-import { sprintf, __ } from '@wordpress/i18n';
```
Nice catch!

techanvil commented:
Hey @eugene-manuilov, I have spotted something else, which is a bit odd.

I tried running a build for this PR and examining the size. I then commented out the `plugins: [ 'lodash' ]` line in `webpack.config.js` and ran the build again - this time, the build was actually smaller. It seems something is not right here - could you please take a look at this?

```sh
npm run build
npm run zip
```

With `plugins: [ 'lodash' ]`:
![image](https://user-images.githubusercontent.com/18395600/220707633-082d2db5-1a8b-436d-999c-e4ccbc47a7e6.png)

Without `plugins: [ 'lodash' ]`:
![image](https://user-images.githubusercontent.com/18395600/220707833-0a365f54-7c90-4607-826b-eee9ac1d71ac.png)

eugene-manuilov commented:
> Nice work here @eugene-manuilov. There are a couple of `import`s remaining that should be updated for consistency:
> 
> https://github.com/google/site-kit-wp/blob/d72e70ed50281c661629e42b636e970f34718ab1/.storybook/utils/resetGlobals.js#L22
> 
> https://github.com/google/site-kit-wp/blob/d72e70ed50281c661629e42b636e970f34718ab1/tests/js/utils.js#L4-L5

Thanks, @techanvil, but those files aren't used in the codebase and not included into the bundle. So changing anything there will have no effect because we use the lodash plugin only for those scripts that are imported in the codebase.

techanvil commented:
> Thanks, @techanvil, but those files aren't used in the codebase and not included into the bundle. So changing anything there will have no effect because we use the lodash plugin only for those scripts that are imported in the codebase.

Thanks @eugene-manuilov, I do appreciate that point, however the AC and IB both mention updating all imports throughout the codebase, and it seems reasonable to update these as well for the sake of consistency...

eugene-manuilov commented:
> 

Ok, updated.



> Hey @eugene-manuilov, I have spotted something else, which is a bit odd.
> 
> I tried running a build for this PR and examining the size. I then commented out the `plugins: [ 'lodash' ]` line in `webpack.config.js` and ran the build again - this time, the build was actually smaller. It seems something is not right here - could you please take a look at this?
> 
> ```shell
> npm run build
> npm run zip
> ```
> 
> With `plugins: [ 'lodash' ]`: ![image](https://user-images.githubusercontent.com/18395600/220707633-082d2db5-1a8b-436d-999c-e4ccbc47a7e6.png)
> 
> Without `plugins: [ 'lodash' ]`: ![image](https://user-images.githubusercontent.com/18395600/220707833-0a365f54-7c90-4607-826b-eee9ac1d71ac.png)

Nice catch, @techanvil. This is weird but looks like webpack itself optimises lodash imports much better than when we use that babel plugin. I ran bundle analyzer and it showed to me that if we used that plugin, the vendors bundle contained lodash files imported as separate modules. Without that plugin, webpack concatenated all lodash files into one module. Maybe we should remove the `babel-plugin-lodash` plugin then?

techanvil commented:
> Ok, updated remaining files as well.

Cheers :+1: 

> Nice catch, @techanvil. This is weird but looks like webpack itself optimises lodash imports much better than when we use that babel plugin. I ran bundle analyzer and it showed to me that if we used that plugin, the vendors bundle contained lodash files imported as separate modules. Without that plugin, webpack concatenated all lodash files into one module. Maybe we should remove the `babel-plugin-lodash` plugin then?

That's interesting! Yeah, it sounds like we should remove `babel-plugin-lodash`, obviously with a sanity check to make sure there's no hidden dependency on it, and the build still builds correctly with it removed...

eugene-manuilov commented:
> > Ok, updated remaining files as well.
> 
> Cheers 👍
> 
> > Nice catch, @techanvil. This is weird but looks like webpack itself optimises lodash imports much better than when we use that babel plugin. I ran bundle analyzer and it showed to me that if we used that plugin, the vendors bundle contained lodash files imported as separate modules. Without that plugin, webpack concatenated all lodash files into one module. Maybe we should remove the `babel-plugin-lodash` plugin then?
> 
> That's interesting! Yeah, it sounds like we should remove `babel-plugin-lodash`, obviously with a sanity check to make sure there's no hidden dependency on it, and the build still builds correctly with it removed...

Deleted.

techanvil reviewed (approved):
LGTM, nice one @eugene-manuilov!
